source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -306,6 +306,11 @@ bool SslSocketInfo::peerCertificatePresented() const {
   return cert != nullptr;
 }
 
+bool SslSocketInfo::localCertificatePresented() const {
+  X509* cert = SSL_get_certificate(ssl_.get());
```

Comment:
IIUC this just verifies existence of a client cert and not an acknowledgement of it's verification by the server?

Reply:
I don't think there's a way to know whether server verified, you know whether the handshake is successful or not, that doesn't indicate server verified though. I might be wrong, cc @PiotrSikora

Reply:
@nrjpoddar is correct, the code in this PR (as of right now) only verifies that the TLS client certificate is configured, and not that the established connection is mutually authenticated.

We could potentially install `SSL_CTX_set_cert_cb` and call `SSL_get_client_CA_list` to check if peer requested local certificate, but I'm not aware of any API that allows us to check this directly.

Reply:
That actually means that this is not a new problem. When we report mTLS using `peerCertificatePresented` on the server, that does not mean that the certificate was validated by the server. So at least in istio context, we should make it clear that mTLS attributes are _necessary_ but not sufficient for true mTLS. For telemetry purposes, this is fine, but for policy, it's not safe to rely on peerCertificatePresented or localCertificatePresented. It's entirely possible that both parties present certs but do no validation.

Anyways, I think this PR is still useful for upstream mTLS bit.

Reply:
Totally agree, minor nit the name for this API should be `localCertificateConfigured` for correctness.

Reply:
@kyessenov actually, it's a new problem.

`peerCertificatePresented` at the server means that it saw a client certificate, which implies that server asked or it. Envoy doesn't currently have an option to request but not verify client certificate (although, there is a PR in flight to allow that), so `peerCertificatePresented` implies that the established connection is mutually authenticated, whereas `localCertificatePresented` doesn't  (unless, I've missed something obvious).

If we want to use `localCertificatePresented` as an indicator of whether or not the connection is mutually authenticated, when acting as a client, then we should use `SSL_CTX_set_cert_cb` and `SSL_get_client_CA_list` to verify that the server requested client certificate and that we sent one on the response, otherwise this field will be misleading.

cc @davidben to see if there is any better way to verify that.

Reply:
The fact that a server requests a client certificate does not mean that:
- the server verified it (in general, e.g. you mention a new change coming), or that
- the client verified the server cert

So even on the server, just checking that the peer presented a cert does not imply that the connection is mutually authenticated.

I agree, that calling it `mtls` is just confusing on both sides. Unless there is a way to transfer the bit saying that the connection is mutually verified over the connection, it doesn't seem to be possible to detect mTLS reliably.

Reply:
In general, yes... but in the current Envoy implementation, server requesting client certificate and not failing the handshake means that it verified it.

However, you're right regarding the client-side verification.

In general, I don't think it's possible to verify that both peers verified the connection from the point of a view of a single peer, but we could verify that client verified identity of the server, and that server verified identity of the client... we'd just need to merge this information somewhere.

Furthermore, in Istio/CSM, we should know the configuration of both peers, so we should know if the configuration requires peer verification or not, shouldn't we?

Reply:
Verification itself is actually a different topic altogether. Client could be checking for a variety of attributes (SAN membership, subject, root cert chain) and server does even less often (only root cert chain). So calling all of that mTLS is also confusing. The common denominator is that both parties present certificates whether they are requested or not. I think that's a good enough approximation to observe the effect of the policy rollout.

Regarding Istio configuration, yes, but the telemetry should not depend on the config rollout process. This is actually what should signal that policies have not rolled out yet fully. We could reflectively analyze listener and cluster configs to detect whether the checks there are sufficient to determine mTLS status. But it seems more reliable to get that information from "what happened" instead of "what should have happened".

Reply:
Regardless, I think that we could expose indicators (i.e. `local_certificate_requested`, `peer_certificate_verified`), and use that, together with deployment-specific context, to figure out whether the connection is mutually authenticated or not out-of-band.

Would that work?

Reply:
Ok, let me update with more indicators.

Reply:
>  The common denominator is that both parties present certificates whether they are requested or not.

TLS client certificates are not presented unless requested by the server.

Reply:
Updated to record server requesting client certs from a callback. I left out CA check since server could use hash verification alone.

Reply:
I don't believe we store this information in BoringSSL right now. I haven't looked at the PR, but of the two suggested options (`SSL_CTX_set_cert_cb` and `SSL_get_client_CA_list`), we only promise that `SSL_get_client_CA_list` works inside the callback anyway, so that option doesn't make sense. (To save memory, we drop the CA list after the handshake completes.)

The `SSL_CTX_set_cert_cb` callback will tell you if the server sent a CertificateRequest. This is not *quite* the same as whether the connection is mutually authenticated in the case of session resumption. Servers typically to carry over client certificate authentication on resumption, since resumption doesn't get to present new certificates. BoringSSL, of course, won't call the callback in that case because there's no certificate to present, and we don't remember in the session that we'd presented a certificate. You would need to track that yourself, if you care about that.

Reply:
It has been great learning just watching the thread. We are mixing few different things here, that I hope to clarify:

First, how can I tell a particular connection is mTLS reliably from a perspective of one peer i.e. client or server. I don't think this is always possible but can be closely approximated by this logic.

* client reports mTLS true if it verified server certificate, presented local certificate & handshake was successful
* server reports mTLS true if it presented local certificate, verified client certificate & handshake was successful
* If both side reports mTLS false it is definitely plain text
* If one side reports false and other true, we cannot definitely say if the connection in mTLS, TLS or plaintext

The second issue we are trying to solve here is get enough information via Envoy to figure out a config distribution problem i.e. control plane says use mTLS but the client/server never presents a certificate.  In this scenario exposing the individual bits of information on a peer like requested peer certificate, presented local certificate & verified peer certificate helps.

The SSL session resumption afaik should just carry forward the mTLS boolean and other information bits that is stored the first time the handshake was finished.

Reply:
Neeraj, agree with the mTLS detection assessment.

Reply:
@davidben I suggested to install callback using `SSL_CTX_set_cert_cb` AND use `SSL_get_client_CA_list` inside the callback to detect if server requested client certificate that matches issuer of the configured client certificate (although, we don't do that currently, so it might be an overkill for this PR), i.e. it was a single suggestion, not two. Sorry if that wasn't clear enough.

Regarding session resumption, you are, of course, right (and thank you for the reminder, I keep forgetting about this more often that I should have)...

For clients, we could store this bit (and possibly others) when adding session keys to the client-side cache in Envoy, and then report value used to establish the original connection.

For servers, we rely on BoringSSL for caching, and we don't store anything in Envoy, so that would be tricky... but we include all configuration related to the client certificate verification in `SSL_CTX_set_session_id_context`, so sessions cannot be resumed across different configurations, which means that we can infer mTLS bit from the successful handshake and the server-side configuration alone.

Reply:
What would be the benefit of checking `SSL_get_client_CA_list`? BoringSSL sends the client certificate, if configured + requested, independent of `SSL_get_client_CA_list`. The CA list is purely a hint. On mismatch, likely the handshake will fail (in which case you don't care). If it succeeds anyway, then the server accepted your client certificate. (Or maybe it silently ignored it, but it's always possible for the server to silently ignore it, which gets to the question of what you're even measuring.)

> For servers, we rely on BoringSSL for caching, and we don't store anything in Envoy, so that would be tricky

This is easy for servers. Just check if `SSL_get_peer_certificate` returns anything. We store the *peer* certificate in the session because that's a critical part of authentication. We don't store the local certificate because that's a waste of memory and ticket size. But, as a server, you always send a local certificate. It's the client certificate that's optional.

Reply:
Let's not check for SSL_get_client_CA_list. That does not handle hash verification on the server-side. I think checking that a client certificate was requested is sufficient for now.